system/lib/wasmfs/syscalls.cpp

```diff
@@ -879,35 +808,24 @@ long __syscall_rename(long old_path, long new_path) {
 
   auto oldPathParts = splitPath((char*)old_path);
 
-  if (oldPathParts.empty()) {
-    return -EINVAL;
-  }
-
-  // In Linux, renaming the root directory returns EBUSY.
-  // TODO: Fix this when path parsing is refactored.
-  std::vector<std::string> root = {"/"};
-  if (oldPathParts == root) {
-    return -EBUSY;
-  }
-
-  auto oldBase = oldPathParts.back();
-
+  // For this operation to be atomic we must lock the source parent directory.
+  // getParsedPath() will return a locked parent directory.
   long err;
-  auto oldParentDir = getDir(oldPathParts.begin(), oldPathParts.end() - 1, err);
+  auto oldParsedPath = getParsedPath(oldPathParts, err);
 
-  if (!oldParentDir) {
+  if (!oldParsedPath.parent) {
     return err;
   }
 
-  // For this operation to be atomic we must lock the source parent directory.
-  auto lockedOldParentDir = oldParentDir->locked();
-
-  auto oldPath = lockedOldParentDir.getEntry(oldBase);
-
-  if (!oldPath) {
+  if (!oldParsedPath.child) {
     return -ENOENT;
   }
 
+  // This matches JS file system behavior.
```

Comment:
It would be good to expand this comment with what the native FS behavior would be as well.